File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3026_maximum-good-subarray-sum/solutions/python3/solution.py
class Solution:
    def maximumSubarraySum(self, nums: list, k: int) -> int:
        from collections import defaultdict
        
        # Map to store sums with the keys as the target value (nums[i] - k and nums[i] + k)
        value_to_indices = defaultdict(list)
        max_sum = float('-inf')
        current_sum = 0

        # Process each number in the array
        for i, num in enumerate(nums):
            current_sum += num
            # Check for good subarray starting point by matching with nums[i] - k and nums[i] + k
            target1 = num - k
            target2 = num + k
            if target1 in value_to_indices:
                for sum_val in value_to_indices[target1]:
                    max_sum = max(max_sum, current_sum - sum_val)
            if target2 in value_to_indices:
                for sum_val in value_to_indices[target2]:
                    max_sum = max(max_sum, current_sum - sum_val)
            # Store current sum in map for future reference with this as the ending point
            value_to_indices[num].append(current_sum)

        return max_sum if max_sum != float('-inf') else 0

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3026_maximum-good-subarray-sum/solutions/java/solution.java
import java.util.HashMap;

class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        HashMap<Integer, Long> map = new HashMap<>();
        long maxSum = 0;
        boolean found = false;
        
        for (int i = 0; i < nums.length; i++) {
            int target1 = nums[i] + k;
            int target2 = nums[i] - k;
            long currentSum = nums[i];
            
            if (map.containsKey(target1)) {
                long newSum = map.get(target1) + nums[i];
                if (newSum > maxSum) {
                    maxSum = newSum;
                    found = true;
                }
            }
            
            if (map.containsKey(target2)) {
                long newSum = map.get(target2) + nums[i];
                if (newSum > maxSum) {
                    maxSum = newSum;
                    found = true;
                }
            }
            
            map.put(nums[i], Math.max(map.getOrDefault(nums[i], 0L), currentSum));
        }
        
        return found ? maxSum : 0;
    }
}
